tmk_core/common/chibios/eeprom_stm32f4.c

```diff
@@ -0,0 +1,378 @@
+    if (status == FLASH_COMPLETE) {
+        /* if the previous operation is completed, proceed to erase the page */
+        FLASH->CR &= ~FLASH_CR_SNB;
+        FLASH->CR |= (sector << FLASH_CR_SNB_POS);
+        FLASH->CR |= FLASH_CR_SER;
+        FLASH->CR |= FLASH_CR_STRT;
+
+        /* Wait for last operation to be completed */
+        status = FLASH_WaitF4();
+        /* clear the erase bit */
+        FLASH->CR &= ~FLASH_CR_SNB;
+        FLASH->CR &= ~FLASH_CR_SER;
+        if (status != FLASH_COMPLETE) {
+            FLASH_ClearFlagF4();
+        }
+    }
+    /* Return the Erase Status */
+    return status;
+}
+
+static FLASH_Status FLASH_ProgramWordF4(uint32_t address, uint32_t data) {
+    FLASH_Status status;
+    if ((address % sizeof(uint32_t) != 0) || !IS_VALID_FLASH(address)) {
+        // not a valid address
+        return FLASH_BAD_ADDRESS;
+    }
+
+    /* Wait for last operation to be completed */
+    status = FLASH_WaitF4();
+    if (status == FLASH_COMPLETE) {
+        // set size to word
+        FLASH->CR &= ~FLASH_CR_PSIZE;
+        FLASH->CR |= FLASH_PSIZE_WORD;
+        FLASH->CR |= FLASH_CR_PG;
+
+        *(__IO uint32_t *)address = data;
+        /* Wait for last operation to be completed */
+        status = FLASH_WaitF4();
+        /* clear the pg bit */
+        FLASH->CR &= ~FLASH_CR_PG;
+        if (status != FLASH_COMPLETE) {
+            FLASH_ClearFlagF4();
+        }
+    }
+    return status;
+}
+
+static uint32_t FLASH_FindValidAddress(void) {
+    uint32_t begin = EE_SECTOR_BEGIN;
+    uint32_t end = EE_SECTOR_END;
+    while( begin < end) {
+        uint32_t data = *(__IO uint32_t*)(begin);
+        if (data == FLASH_EMPTY_VALUE) {
+            return begin;
+        }
+        begin += 4;
+    }
+    return FLASH_EMPTY_VALUE;
+}
+
+static void FLASH_Backup(void) {
+    uint32_t begin = EE_SECTOR_BEGIN;
+    uint32_t end = EE_SECTOR_END;
+    memset(&DataBuf[0], 0xFF, sizeof(DataBuf));
+    while( begin < end) {
+        uint32_t data = *(__IO uint32_t*)(begin);
+        if (IS_VALID_ADDRESS(DATA_ADDRESS(data))){
+            DataBuf[DATA_ADDRESS(data)] = DATA_VALUE(data);
+        }
+        begin += 4;
+    }
+}
+
+static uint16_t FLASH_Restore(void) {
+    uint32_t current = EE_SECTOR_BEGIN;
+    for (uint8_t i = 0; i < EE_MAX_BYTES; i++) {
+        if (DataBuf[i] != 0xFF) {
+            uint32_t data = DATA_COMPOSE(i, DataBuf[i]);
+            FLASH_ProgramWordF4(current, data);
+            current += 4;
+        }
+    }
+    return FLASH_COMPLETE;
+}
+
+/*******************************************************************
+ * stm32 eeprom interface
+ *******************************************************************/
+uint16_t EEPROM_Init(void) {
+    FLASH_UnlockF4();
+    return FEE_DENSITY_BYTES;
+}
+
+void EEPROM_Erase(void) {
+    FLASH_EraseSectorF4(EE_SECTOR_ID);
+}
+
+uint16_t EEPROM_WriteDataByte(uint16_t Address, uint8_t DataByte) {
+    uint32_t addr = FLASH_FindValidAddress();
```

Comment:
Given there's a possibility of using the "shadow" copy in the static `DataBuf` array, to increase the endurance perhaps use it as a cache and compare against the existing value before attempting a write. Writing the same value to the same address will needlessly add entries to flash with this implementation.
Reads can follow the same principle, consulting the shadow copy, rather than iterating through the flash?